File: current_session/1249.py
"""
see https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/
"""
class Solution(object):
    def minRemoveToMakeValid(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = []
        res = []
        for i in range(len(s)):
            if s[i] in '()':
                if s[i] == '(':
                    stack += i,
                else:
                    if stack:
                        res += stack.pop(),
                        res += i,
            else:
                res += i,
        res.sort()
        return ''.join([s[i] for i in res])

print(Solution().minRemoveToMakeValid("lee(t(c)o)de)"))
print(Solution().minRemoveToMakeValid("a)b(c)d"))
print(Solution().minRemoveToMakeValid("))(("))
print(Solution().minRemoveToMakeValid("(a(b(c)d)"))

File: current_session/c++/1249.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

class Solution {
public:
  string minRemoveToMakeValid(string s) {
    stack<int> st;
    // use stack, mark the invalid char as '#'
    for (int i=0; i<s.size(); ++i) {
      if (s[i] == '(') {
        st.push(i);
      } else if (s[i] == ')') {
        if (!st.empty()) {
          st.pop();
        } else {
          s[i] = '#';
        }
      }
    }
    while (!st.empty()) {
      s[st.top()] = '#';
      st.pop();
    }
    s.erase(remove(s.begin(), s.end(), '#'), s.end());
    return s;
  }
};